### PR TITLE
Aarch64 transition to EL1

### PR DESCRIPTION
**Pull request by gmacd**
- (Boot) Transition to EL1 on startup
- (qemu) Show exception level change events in stdout
- (qemu) Add option to start qemu in gdb mode and wait for connection

There'll be more to do as we set up interrupts and timers, but this will do for now.

Signed-off-by: Graham MacDonald <grahamamacdonald@gmail.com>

### CODE REVIEW

**Review by dancrossnyc — APPROVED** [review]

**Comment by gmacd** [comment]
Fixed build - could I get another review @dancrossnyc ?  Thanks

File: aarch64/src/devcons.rs [hunk 1]
```diff
@@ -60,8 +60,8 @@ const MBOX_EMPTY: u32 = 0x4000_0000;
 
 // Delay for count cycles
 fn delay(count: u32) {
-    #[cfg(not(test))]
     for _ in 0..count {
+        #[cfg(not(test))]
```
Comment (dancrossnyc) [inline aarch64/src/devcons.rs:64]: Actually, I don't see any reason not to remove this entirely.

File: aarch64/src/kernel.ld [hunk 2]
```diff
@@ -6,9 +6,7 @@ ENTRY(start)
 
 SECTIONS {
 	/*
-	 * start the kernel at 0x0xffff_8000_0020_0000
-	 * This preserves some RAM between 1MiB and the
-	 * start of the kernel for critical structures.
+	 * Raspberry Pi start address is 0x80_0000.
```
Comment (dancrossnyc) [inline aarch64/src/kernel.ld:9]: Whoa. It really, really shouldn't be. I'd leave the start address here the same as it is on x86.

It's true right now we're running down at low addresses, but that's not tenable long-term.

File: aarch64/src/l.S [hunk 3]
```diff
@@ -2,15 +2,82 @@
 
 STACKSZ = 4096*4
 
+.equ	CURRENTEL_EL,		(1<<3) | (1<<2)
```
Comment (dancrossnyc) [inline aarch64/src/l.S:5]: Don't use `.equ`; just use `=` for consistency with the rest of the code.

File: aarch64/src/l.S [hunk 4]
```diff
@@ -2,15 +2,82 @@
...
+
 .section .boottext, "awx"
 .globl start
 start:
 	// All cores other than 0 should just hang
 	mrs	x0, mpidr_el1
 	and	x0, x0, #0xff
 	cbnz	x0, dnr
+
+	// Aarch64 has 4 exception levels:
```
Comment (dancrossnyc) [inline aarch64/src/l.S:37]: Exception or execution?

File: aarch64/src/l.S [hunk 5]
```diff
@@ -2,15 +2,82 @@
...
+	mrs	x0, CurrentEL
+	and	x0, x0, CURRENTEL_EL
+	lsr	x0, x0, #2
+	cmp	x0, #1
+	beq	el1
+	cmp	x0, #2
+	beq	el2
+
+	// Must be EL3, so prepare jump to EL2
+	ldr	x0, =(SCR_EL3_NS|SCR_EL3_SMD|SCR_EL3_HCE|SCR_EL3_RW)
```
Comment (dancrossnyc) [inline aarch64/src/l.S:54]: If I understand this correctly, this is basically loading an "address" created by OR'ing together a bunch of bits into x0. Why not simply use MOV with a literal in this case?

File: xtask/src/main.rs [hunk 6]
```diff
@@ -37,8 +38,9 @@ impl BuildParams {
             if matches.contains_id("release") { Profile::Release } else { Profile::Debug };
         let verbose = matches.contains_id("verbose");
         let arch = Arch::Aarch64;
+        let wait_for_gdb = matches.try_contains_id("gdb").unwrap_or(false);
```
Comment (dancrossnyc) [inline xtask/src/main.rs:41]: Why do you need this instead of `contains_id`?

File: aarch64/src/devcons.rs (hunk 1 above)
Comment (gmacd) [reply aarch64/src/devcons.rs:64]: yeah, funnily I was wondering the same thing right now

File: aarch64/src/l.S (hunk 4 above)
Comment (gmacd) [reply aarch64/src/l.S:37]: exception
https://developer.arm.com/documentation/102412/0102/Privilege-and-Exception-levels

File: aarch64/src/l.S (hunk 3 above)
Comment (gmacd) [reply aarch64/src/l.S:5]: where are we using `=` in assembly?  I see `.set`, which is apparently mutable, so felt `.equ` would be better.

File: aarch64/src/kernel.ld (hunk 2 above)
Comment (gmacd) [reply aarch64/src/kernel.ld:9]: https://www.raspberrypi.com/documentation/computers/config_txt.html#kernel_address

Should be 0x8000 for 32bit, 0x200000 for 64 bit.  (I had neither - 0x80000 by accident - I don't think qemu is too fussy, within a certain range).

The link also states we should be able to specify the address in `config.txt`.  I'll have to try it out on an actual rpi...

File: aarch64/src/l.S (hunk 3 above)
Comment (dancrossnyc) [reply aarch64/src/l.S:5]: `=` seemed briefer and serves largely the same purpose.

File: aarch64/src/kernel.ld (hunk 2 above)
Comment (dancrossnyc) [reply aarch64/src/kernel.ld:9]: This is conflating two separate things. Those are _physical_ addresses, where the loader puts the kernel binary in RAM, but the linker config file specifies _virtual_ addresses, assuming that we're running the kernel under virtual memory. The kernel should be linked at whatever virtual addresses it expects to run at.

Now, we haven't turned on paging yet, so for now it's ok to link the kernel at a physical address, but it won't run there long-term.

File: xtask/src/main.rs (hunk 6 above)
Comment (gmacd) [reply xtask/src/main.rs:41]: I'm wondering if we should move to a struct for defining the subcommands now that they're diverging: https://docs.rs/clap/3.2.14/clap/_derive/_tutorial/index.html

File: xtask/src/main.rs [hunk 7]
```diff
@@ -37,8 +38,9 @@ impl BuildParams {
             if matches.contains_id("release") { Profile::Release } else { Profile::Debug };
         let verbose = matches.contains_id("verbose");
         let arch = Arch::Aarch64;
```
[the remark this answers is not in the document]
Comment (dancrossnyc) [reply xtask/src/main.rs:40]: Oh, wait; `x86_64` uses `.set`? Sigh. Ok, keep it as-is for now, then, and we'll normalize between the two at some point in the future.

File: xtask/src/main.rs (hunk 6 above)
Comment (dancrossnyc) [reply xtask/src/main.rs:41]: That could be cool; I certainly wouldn't be opposed. I suggest we land this more or less as-is now, and then perhaps circle back to it later.

**Review by dancrossnyc — APPROVED** [review]

**Pull request merged**